server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,440 @@
+        boolean isAutoThrottle = mergeTrigger != MergeTrigger.CLOSING && merge.getStoreMergeInfo().mergeMaxNumSegments() == -1;
+        // IO throttling cannot be toggled for existing merge tasks, only new merge tasks pick up the updated IO throttling setting
+        return new MergeTask(
+            mergeSource,
+            merge,
+            isAutoThrottle && config.isAutoThrottle(),
+            "Lucene Merge Task #" + mergeTaskCount.incrementAndGet() + " for shard " + shardId
+        );
+    }
+
+    // synchronized so that {@code #currentlyRunningMergeTasks} and {@code #backloggedMergeTasks} are modified atomically
+    private synchronized boolean runNowOrBacklog(MergeTask mergeTask) {
+        assert mergeTask.isRunning() == false;
+        if (currentlyRunningMergeTasks.size() >= config.getMaxThreadCount()) {
+            backloggedMergeTasks.add(mergeTask);
+            return false;
+        } else {
+            boolean added = currentlyRunningMergeTasks.add(mergeTask);
+            assert added : "starting merge task [" + mergeTask + "] registered as already running";
+            return true;
+        }
+    }
+
+    private void mergeDone(MergeTask mergeTask) {
+        assert mergeTask.isRunning();
+        synchronized (this) {
+            boolean removed = currentlyRunningMergeTasks.remove(mergeTask);
+            assert removed : "completed merge task [" + mergeTask + "] not registered as running";
+            // when one merge is done, maybe a backlogged one can now execute
+            enqueueBacklogged();
+        }
+        mergeTaskDoneCount.incrementAndGet();
+        checkMergeTaskThrottling();
+    }
+
+    private synchronized void enqueueBacklogged() {
+        int maxBackloggedTasksToEnqueue = config.getMaxThreadCount() - currentlyRunningMergeTasks.size();
+        while (maxBackloggedTasksToEnqueue-- > 0) {
+            MergeTask backloggedMergeTask = backloggedMergeTasks.poll();
+            if (backloggedMergeTask == null) {
+                break;
+            }
+            threadPoolMergeQueue.enqueueMergeTask(backloggedMergeTask);
+        }
+    }
+
+    /**
+     * Does the actual merge, by calling {@link org.apache.lucene.index.MergeScheduler.MergeSource#merge}
+     */
+    protected void doMerge(MergeSource mergeSource, MergePolicy.OneMerge merge) throws IOException {
+        mergeSource.merge(merge);
+    }
+
+    @Override
+    public Directory wrapForMerge(MergePolicy.OneMerge merge, Directory in) {
+        // Return a wrapped Directory which has rate-limited output.
+        // Note: the rate limiter is only per thread. So, if there are multiple merge threads running
+        // and throttling is required, each thread will be throttled independently.
+        // The implication of this, is that the total IO rate could be higher than the target rate.
+        RateLimiter rateLimiter = onGoingMergeRateLimiter.get();
```

Comment:
I wonder if we should key the `currentlyRunningMergeTasks` by the one-merge instead (as a map) to avoid the use of thread-local. I do recognize this is similar to the existing merge scheduler but feels slightly more intuitive to me now that we register the tasks anyway.

Reply:
Great idea! Pushed [6a96a02](https://github.com/elastic/elasticsearch/pull/120869/commits/6a96a027407bac8fc4eb1472e20be3b8b6b7b9bc)

Reply:
It might not work, I'll investigate.

Reply:
> It might not work, I'll investigate.

False alarm [d170de8](https://github.com/elastic/elasticsearch/pull/120869/commits/d170de8e7e74cbd9864985d99b791f996501fe30) . It should be working now.